libnet/common.h

```diff
@@ -0,0 +1,97 @@
+
+#include <semaphore.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#define HEADER_LEN 2
+typedef struct {
+	int fd;
+	short state;
+	int read_status;
+	int retries;
+	unsigned char internal_buffer[HEADER_LEN]; // used to store partially retreived data
+} client_info;
+
+typedef struct {
+	unsigned char tag;
+	sem_t number;
+} available_tag_t;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic warning "-pedantic"
+#pragma message "This is not strictly C99 but works and is awfully convenient"
+typedef struct {
+	unsigned char tag;
+	size_t length;
+	unsigned char value[0]; // variable size array at the tail of struct
+} tlv;
+#pragma GCC diagnostic pop
+
+// as per https://lists.gnu.org/archive/html/bug-gnulib/2006-06/msg00014.html
+#ifndef SSIZE_MAX
+# define SSIZE_MAX ((ssize_t) (SIZE_MAX /2))
+#endif
+
+#define TAG_INTERNAL_MASK 0xe0
+
+#define TAG_HELO 0xe0
+#define TAG_BYE  0xe1
+#define TAG_ECHO 0xe2
+#define TAG_OK   0xe3
+
+#define STATE_NONE 0 // free slot
+#define STATE_CONNECTING 1
+#define STATE_READY 2
+#define STATE_DISCONNECTED 3
+#define STATE_ERROR 4
+
+#define READ_STATUS_DEFAULT 0
+#define READ_STATUS_WAITING_FOR_DATA 1
+
+#ifndef MAX_MESSAGE_QUEUE
+# define MAX_MESSAGE_QUEUE 100
+#endif
+
+tlv** get_message_queue();
+
```

Comment:
This contains only functions used internally in libnet, that provide shared functionality for server and client and is not intended for outside use. 
For client/server interface see `include/client.h` and `include/server.h`.
Will fix if there's time